---
title: ActiveSupport::Rescuable::ClassMethods
layout: default
---
<div class="main">
    <div class="banner">
        
            <span>Ruby on Rails 7.0.2</span><br />
        
        <div class="type">Module</div>
        <h1>
            ActiveSupport::Rescuable::ClassMethods
            
        </h1>
        <ul class="files">
            
            <li><a href="../../../files/activesupport/lib/active_support/rescuable_rb.html">activesupport/lib/active_support/rescuable.rb</a></li>
            
        </ul>
    </div>
    <div id="bodyContent">
        <div id="content">
  

  

  
  


  

  
    <h2 id="methods">Methods</h2>
    <ul>
      
        <li>
          <a href="#method-i-rescue_from">rescue_from</a>
        </li>
      
        <li>
          <a href="#method-i-rescue_with_handler">rescue_with_handler</a>
        </li>
      
    </ul>
  

  

  
    

    

    

    

    <!-- Methods -->
    
    
      <h2 id="instance-public-methods">Instance Public methods</h2>
      
        <div class="method">
          <h3 id="method-i-rescue_from">
            
              rescue_from(*klasses, with: nil, &amp;block)
            
          </h3>

          
            <div class="description">
              <p>Registers exception classes with a handler to be called by <code>rescue_with_handler</code>.</p>

<p><code>rescue_from</code> receives a series of exception classes or class names, and an exception handler specified by a trailing <code>:with</code> option containing the name of a method or a Proc object. Alternatively, a block can be given as the handler.</p>

<p>Handlers that take one argument will be called with the exception, so that the exception can be inspected when dealing with it.</p>

<p>Handlers are inherited. They are searched from right to left, from bottom to top, and up the hierarchy. The handler of the first class for which <code>exception.is_a?(klass)</code> holds true is the one invoked, if any.</p>

<pre><code>class ApplicationController &lt; ActionController::Base
  rescue_from User::NotAuthorized, with: :deny_access # self defined exception
  rescue_from ActiveRecord::RecordInvalid, with: :show_errors

  rescue_from &#39;MyAppError::Base&#39; do |exception|
    render xml: exception, status: 500
  end

  private
    def deny_access
      ...
    end

    def show_errors(exception)
      exception.record.new_record? ? ...
    end
end
</code></pre>

<p>Exceptions raised inside exception handlers are not propagated up.</p>
            </div>
          

          

          

          
            
            <div class="sourcecode">
              
              <p class="source-link">
                Source:
                
              </p>
              <div id="method-i-rescue_from_source">
                <pre><code class="ruby"><span class="ruby-comment"># File activesupport/lib/active_support/rescuable.rb, line 51</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier ruby-title">rescue_from</span>(<span class="ruby-operator">*</span><span class="ruby-identifier">klasses</span>, <span class="ruby-value">with:</span> <span class="ruby-keyword">nil</span>, <span class="ruby-operator">&amp;</span><span class="ruby-identifier">block</span>)
  <span class="ruby-keyword">unless</span> <span class="ruby-identifier">with</span>
    <span class="ruby-keyword">if</span> <span class="ruby-identifier">block_given?</span>
      <span class="ruby-identifier">with</span> = <span class="ruby-identifier">block</span>
    <span class="ruby-keyword">else</span>
      <span class="ruby-identifier">raise</span> <span class="ruby-constant">ArgumentError</span>, <span class="ruby-string">&quot;Need a handler. Pass the with: keyword argument or provide a block.&quot;</span>
    <span class="ruby-keyword">end</span>
  <span class="ruby-keyword">end</span>

  <span class="ruby-identifier">klasses</span>.<span class="ruby-identifier">each</span> <span class="ruby-keyword">do</span> <span class="ruby-operator">|</span><span class="ruby-identifier">klass</span><span class="ruby-operator">|</span>
    <span class="ruby-identifier">key</span> = <span class="ruby-keyword">if</span> <span class="ruby-identifier">klass</span>.<span class="ruby-identifier">is_a?</span>(<span class="ruby-constant">Module</span>) <span class="ruby-operator">&amp;&amp;</span> <span class="ruby-identifier">klass</span>.<span class="ruby-identifier">respond_to?</span>(<span class="ruby-value">:===</span>)
      <span class="ruby-identifier">klass</span>.<span class="ruby-identifier">name</span>
    <span class="ruby-keyword">elsif</span> <span class="ruby-identifier">klass</span>.<span class="ruby-identifier">is_a?</span>(<span class="ruby-constant">String</span>)
      <span class="ruby-identifier">klass</span>
    <span class="ruby-keyword">else</span>
      <span class="ruby-identifier">raise</span> <span class="ruby-constant">ArgumentError</span>, <span class="ruby-node">&quot;#{klass.inspect} must be an Exception class or a String referencing an Exception class&quot;</span>
    <span class="ruby-keyword">end</span>

    <span class="ruby-comment"># Put the new handler at the end because the list is read in reverse.</span>
    <span class="ruby-keyword">self</span>.<span class="ruby-identifier">rescue_handlers</span> <span class="ruby-operator">+=</span> [[<span class="ruby-identifier">key</span>, <span class="ruby-identifier">with</span>]]
  <span class="ruby-keyword">end</span>
<span class="ruby-keyword">end</span></code></pre>
              </div>
            </div>
            
          </div>
        
        <div class="method">
          <h3 id="method-i-rescue_with_handler">
            
              rescue_with_handler(exception, object: self, visited_exceptions: [])
            
          </h3>

          
            <div class="description">
              <p>Matches an exception to a handler based on the exception class.</p>

<p>If no handler matches the exception, check for a handler matching the (optional) exception.cause. If no handler matches the exception or its cause, this returns <code>nil</code>, so you can deal with unhandled exceptions. Be sure to re-raise unhandled exceptions if this is what you expect.</p>

<pre><code>begin
  â€¦
rescue =&gt; exception
  rescue_with_handler(exception) || raise
end
</code></pre>

<p>Returns the exception if it was handled and <code>nil</code> if it was not.</p>
            </div>
          

          

          

          
            
            <div class="sourcecode">
              
              <p class="source-link">
                Source:
                
              </p>
              <div id="method-i-rescue_with_handler_source">
                <pre><code class="ruby"><span class="ruby-comment"># File activesupport/lib/active_support/rescuable.rb, line 88</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier ruby-title">rescue_with_handler</span>(<span class="ruby-identifier">exception</span>, <span class="ruby-value">object:</span> <span class="ruby-keyword">self</span>, <span class="ruby-value">visited_exceptions:</span> [])
  <span class="ruby-identifier">visited_exceptions</span> <span class="ruby-operator">&lt;&lt;</span> <span class="ruby-identifier">exception</span>

  <span class="ruby-keyword">if</span> <span class="ruby-identifier">handler</span> = <span class="ruby-identifier">handler_for_rescue</span>(<span class="ruby-identifier">exception</span>, <span class="ruby-value">object:</span> <span class="ruby-identifier">object</span>)
    <span class="ruby-identifier">handler</span>.<span class="ruby-identifier">call</span> <span class="ruby-identifier">exception</span>
    <span class="ruby-identifier">exception</span>
  <span class="ruby-keyword">elsif</span> <span class="ruby-identifier">exception</span>
    <span class="ruby-keyword">if</span> <span class="ruby-identifier">visited_exceptions</span>.<span class="ruby-identifier">include?</span>(<span class="ruby-identifier">exception</span>.<span class="ruby-identifier">cause</span>)
      <span class="ruby-keyword">nil</span>
    <span class="ruby-keyword">else</span>
      <span class="ruby-identifier">rescue_with_handler</span>(<span class="ruby-identifier">exception</span>.<span class="ruby-identifier">cause</span>, <span class="ruby-value">object:</span> <span class="ruby-identifier">object</span>, <span class="ruby-value">visited_exceptions:</span> <span class="ruby-identifier">visited_exceptions</span>)
    <span class="ruby-keyword">end</span>
  <span class="ruby-keyword">end</span>
<span class="ruby-keyword">end</span></code></pre>
              </div>
            </div>
            
          </div>
        
      
    
  
</div>

    </div>
</div>
